proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,199 @@
+
+Groups can later be used in the future when it's time to deprecate a set of packages. The package author can request to
+ deprecate a `group` by calling a single command on the nuget CLI and all the packages that belong to that group will be marked as `deprecated`.
+
+With this information in place, new set of NuGet APIs / extensions will be needed to allow a user to run the following actions:
+- **Stage** a package for later publishing
+- **Publish an earlier staged release**
+- **Deprecate a group** using the group id 
+- **Discard a staged release** using the group id
+
+Note, that the user is required to be the owner / co-owner of the packages which are in the group,
+otherwise the request to make any modification to a non-owned package will fail.
+
+#### Package staging
+When a package is staged, the id of the package is reserved at that point, and no other user can introduce a package with the same name.
+
+Also, to avoid abuse, only a limited number of packages can be staged within a single group. While this can potentially be defined as a configurable value per organization / owner, there is no such requirement at the moment. Hence, this can be set to 1000 packages for now or not defined at all. Will leave this for potentially future consideration.
+
+When staged, packages will not be visible / accessible to anyone but the owner (including coowners). They will be able to see the current staged list for a given group using `nuget stage --list --group-id [group-id]` command. 
+
+
+#### Group Ownership
+There are few interesting scenarios related to package ownership:
+1. The user has staged a package to a group for later publishing, but the ownership has changed, and the user is no longer
+ an owner for a package in a group. At this point, publishing request should fail with a message that the group has been modified
+ from its original set, and the outcome of the request may be different from what the user was expecting, indicating the package
+ that was moved out of the group. If the user is aware of the problem and still wants to move forward with the change,
+ they will have to pass an additional `--force` parameter to the publish command: `nuget publish <group-id> --force`.
+2. The group to be published no longer has any packages in it. In this case the command should fail detailing the reason for the failure.
+3. This will be the same but for deprecation. Essentially, any bulk operation using groups, where the set of the packages in the group
+ have changed after staging, should result in an error with details about the change. The `--force` flag should again be used
+ to ignore the change and move forward with execution of the requst (either publishing or deprecation). To summarize,
+ the fundamental principle is that a user cannot deprecate a package they don't own.
+
+#### Lifespan of a staged group
```

Comment:
To simplify this, could there be a set timeline for this feature. Say 1 month? After which these packages just get deleted if in staged mode?

Reply:
I think that's what I have specificed already, haven't I:

> there should be a maximum lifespan for a staged package, that is not published. NuGet should then unstage / remove any package that was staged but never published within the specified / pre-configured time period. As a maximum timeslot a default one month period can be considered.

Maybe I'm misunderstanding you. Or do you mean to not even give the users ability to specify any expiration period, and only stick with the "maximum" one from the server?

Reply:
yes. For MVP, a default expiration should work for all.

Reply:
Got it. I think it will make sense to explicitly call out that the "customized expiration" should be treated as a P2 for the eng. team to consider when they can. I'll make a note of that.